src/component_selector.jl

```diff
@@ -0,0 +1,295 @@
+) = NameComponentSelector(component_subtype, component_name, name)
+"""
+Make a `ComponentSelector` from a component reference, pointing to components in any
+collection with the given component's subtype and name. Optionally provide a name for the
+`ComponentSelector`.
+"""
+make_selector(
+    component_ref::InfrastructureSystemsComponent;
+    name::Union{String, Nothing} = nothing,
+) =
+    make_selector(
+        typeof(component_ref),
+        get_name(component_ref)::AbstractString;
+        name = name,
+    )
+
+# Naming
+default_name(e::NameComponentSelector) =
+    component_to_qualified_string(e.component_subtype, e.component_name)
+
+# Contents
+function get_components(e::NameComponentSelector, sys::SYSTEM_LIKE; filterby = nothing)
+    com = get_component(e.component_subtype, sys, e.component_name)
+    (!isnothing(filterby) && !filterby(com)) && (com = nothing)
+    return (com === nothing) ? [] : [com]
+end
+
+# ListComponentSelector
+"`PluralComponentSelector` represented by a list of other `ComponentSelector`s."
+struct ListComponentSelector <: PluralComponentSelector
+    # Using tuples internally for immutability => `==` is automatically well-behaved
+    content::Tuple{Vararg{ComponentSelector}}
+    name::Union{String, Nothing}
+end
+
+# Construction
+"""
+Make a `ComponentSelector` pointing to a list of sub-selectors, which form the groups.
+Optionally provide a name for the `ComponentSelector`.
+"""
+make_selector(content::ComponentSelector...; name::Union{String, Nothing} = nothing) =
+    ListComponentSelector(content, name)
+
+# Naming
+default_name(e::ListComponentSelector) = "[$(join(get_name.(e.content), ", "))]"
+
+# Contents
+function get_groups(e::ListComponentSelector, sys; filterby = nothing)
+    return e.content
+end
+
+function get_components(e::ListComponentSelector, sys::SYSTEM_LIKE; filterby = nothing)
+    sub_components =
+        Iterators.map(x -> get_components(x, sys; filterby = filterby), e.content)
+    return Iterators.flatten(sub_components)
+end
+
+# SubtypeComponentSelector
+"`PluralComponentSelector` represented by a type of component."
+struct SubtypeComponentSelector <: DynamicallyGroupedComponentSelector
+    component_subtype::Type{<:InfrastructureSystemsComponent}
+    name::Union{String, Nothing}
+    groupby::Union{Symbol, Function}
+end
+
+# Construction
+"""
+Make a `ComponentSelector` from a type of component. Optionally provide a name and/or
+grouping behavior for the `ComponentSelector`.
+"""
+make_selector(
+    component_subtype::Type{<:InfrastructureSystemsComponent};
+    name::Union{String, Nothing} = nothing,
+    groupby::Union{Symbol, Function} = :all,
+) =
+    SubtypeComponentSelector(component_subtype, name, validate_groupby(groupby))
+
+# Naming
+default_name(e::SubtypeComponentSelector) = subtype_to_string(e.component_subtype)
+
+# Contents
+function get_components(e::SubtypeComponentSelector, sys::SYSTEM_LIKE; filterby = nothing)
```

Comment:
So there's this analogy:

- Eager: call `get_components(type, system)`. Lazy: call `make_selector(type)` and later call `get_components` on it.
- Eager: call `get_component(type, system, name)`. Lazy: call `make_selector(type, name)` and later call `get_component` on it.
- Eager: call `get_components(filter_func, type, system)`. Lazy: call `make_selector(type, filter_fn)` and later call `get_components` on it (granted this means I should probably reorder the parameters to that `make_selector` method).

so in that sense, my `filter_fn` below is analogous to the existing `filter_func` and this `filterby` kwarg is something else, just an extra feature to satisfy a special case (I talk about this in the demo notebook).

I'll rename `filter_fn` to `filter_func`. My rationale for giving `filter_fn` and `filterby` different names was that they *are* different things, but maybe that's not a good argument, thoughts?

I don't really foresee users wanting to use a `do` block with this, and I think the fact that `filterby` is an optional add-on to the existing selector interface justifies its inclusion as a kwarg rather than the first parameter, but I don't think it hurts to do both; I can add a wrapper method where the first parameter is `filterby`.

Reply:
Update, I think every part of my interface that takes a `filterby` or `filter_fn` (now `filter_func`) now has the option of putting that first.

Reply:
I've renamed `filterby` to `scope_limiter` to reduce confusion with `filter_func`. The mental model is maybe like a [universe](https://en.wikipedia.org/wiki/Universe_(mathematics)) in math: it can sometimes make sense to limit the scope of objects you are considering from literally everything to some collection that makes sense in the context (e.g., only `available` components) before doing the actually interesting filtering. Open to alternate naming suggestions, also considered `domain_limiter`.

Reply:
Resolving this as `scope_limiter` is no longer part of the user interface.